pvlib/iotools/pvgis.py

```diff
@@ -23,6 +23,353 @@
+    if peakpower is not None:
+        params['peakpower'] = peakpower
+    if loss is not None:
+        params['loss'] = loss
+    if optimal_surface_tilt:
+        params['optimalinclination'] = 1
+    if optimalangles:
+        params['optimalangles'] = 1
+
+    # The url endpoint for hourly radiation is 'seriescalc'
+    res = requests.get(url + 'seriescalc', params=params, timeout=timeout)
+
+    # PVGIS returns really well formatted error messages in JSON for HTTP/1.1
+    # 400 BAD REQUEST so try to return that if possible, otherwise raise the
+    # HTTP/1.1 error caught by requests
+    if not res.ok:
+        try:
+            err_msg = res.json()
+        except Exception:
+            res.raise_for_status()
+        else:
+            raise requests.HTTPError(err_msg['message'])
+
+    # initialize data to None in case API fails to respond to bad outputformat
+    data = None, None, None, None
+    if outputformat == 'json':
+        src = res.json()
+        return _parse_pvgis_hourly_json(src, map_variables=map_variables)
+    elif outputformat == 'csv':
+        with io.StringIO(res.content.decode('utf-8')) as src:
+            return _parse_pvgis_hourly_csv(src, map_variables=map_variables)
+    else:
+        # this line is never reached because if outputformat is not valid then
+        # the response is HTTP/1.1 400 BAD REQUEST which is handled earlier
+        pass
+    return data
+
+
+def _parse_pvgis_hourly_json(src, map_variables):
+    inputs = src['inputs']
+    metadata = src['meta']
+    data = pd.DataFrame(src['outputs']['hourly'])
+    data.index = pd.to_datetime(data['time'], format='%Y%m%d:%H%M', utc=True)
+    data = data.drop('time', axis=1)
+    data = data.astype(dtype={'Int': 'int'})  # The 'Int' column to be integer
+    if map_variables:
+        data.rename(columns=VARIABLE_MAP, inplace=True)
+    return data, inputs, metadata
+
+
+def _parse_pvgis_hourly_csv(src, map_variables):
+    # The first 4 rows are latitude, longitude, elevation, radiation database
+    inputs = {}
+    # 'Latitude (decimal degrees): 45.000\r\n'
+    inputs['latitude'] = float(src.readline().split(':')[1])
+    # 'Longitude (decimal degrees): 8.000\r\n'
+    inputs['longitude'] = float(src.readline().split(':')[1])
+    # Elevation (m): 1389.0\r\n
+    inputs['elevation'] = float(src.readline().split(':')[1])
+    # 'Radiation database: \tPVGIS-SARAH\r\n'
+    inputs['radiation_database'] = str(src.readline().split(':')[1]
+                                       .replace('\t', '').replace('\n', ''))
+    # Parse through the remaining metadata section (the number of lines for
+    # this section depends on the requested parameters)
+    while True:
+        line = src.readline()
+        if line.startswith('time,'):  # The data header starts with 'time,'
+            # The last line of the metadata section contains the column names
+            names = line.replace('\n', '').replace('\r', '').split(',')
```

Comment:
Can the `.replace('\n', '').replace('\r', '')` pattern used here and elsewhere be replaced with `.strip()`?

Reply:
Nice solution. Seems to work fine.